### ISSUE TITLE
Wrong warning that a function is unused (Wunused-function)

### ISSUE THREAD

**Issue by gruenich** [llvm/llvm-project#124469]
I am using Clang to compile a C project. It warns that one of the functions is not used (Wunused-function). When I delete the function, it does not compile anymore, because the function is missing. Looks like the warning is a false positive, it should not be emitted.

```
./../Common/assemble_bndry.c:71:28: warning: unused function 'type_index' [-Wunused-function]
   71 | static inline unsigned int type_index(MATENT_TYPE dsttype,
      |                            ^~~~~~~~~~
1 warning generated.
```
This is LLVM/Clang main from mid January and also occurs in Clang 19.

The project can be found at: https://gitlab.com/alberta-fem/alberta3 and it is based on GNU Autotools. What can I provide, to help debugging? Adding `-E` to my compile command did not work (or I did it wrong). The command generating the warning is:
```
/bin/sh ../../../libtool  --tag=CC   --mode=compile /home/gruenich/llvm-project/build/bin/clang -DHAVE_CONFIG_H -I. -I../../..  -Wall -Wpedantic -I../../../include -I./../0d -I./../1d -I./../2d -I./../3d -I./../Common -I../../../alberta_util/src -I../../../alberta_util/src -I../Common -I/usr/local/include      -DDIM_OF_WORLD=1 -DALBERTA_DEBUG=0 -DALBERTA_PROFILE=0 -DHAVE_ROW_FCTS_V_TYPE=1 -DHAVE_COL_FCTS_V_TYPE=1 -DHAVE_ROW_FCTS_C_TYPE=0 -DHAVE_COL_FCTS_C_TYPE=0 -Wall -Wpedantic -Wall -Wpedantic  -MT ../Common/libVV_assemble_1d_la-assemble_bndry.lo -MD -MP -MF ../Common/.deps/libVV_assemble_1d_la-assemble_bndry.Tpo -c -o ../Common/libVV_assemble_1d_la-assemble_bndry.lo `test -f '../Common/assemble_bndry.c' || echo './'`../Common/assemble_bndry.c
```

**Comment by timmeywerk** [comment]
If you have the time to wait for a newby, I would like to take this one.

**Comment by thesamesam** [comment]
> What can I provide, to help debugging? 

Please include preprocessed sources (`-E` should create a `.i` file).

**Comment by gruenich** [comment]
@timmeywerk I am in no rush. Give it a try!

@thesamesam Thanks for the tip with `-save-temps`, it work while `-E` was ignored.

[assemble_bndry.i.txt](https://github.com/user-attachments/files/18605405/assemble_bndry.i.txt)

**Comment by AaronBallman** [comment]
I can't reproduce the issue: https://godbolt.org/z/o3fY3rsen and the function does indeed seem to be unused (there's only one instance of `type_index` when I search through the file).

**Comment by gruenich** [comment]
I found the issue, it is me or inside the code. I did not fully understand the complicated C macros. I moved `type_index` into one of the macros and now it both compiles and no warnings are emitted.

Thanks for your support and sorry for claiming it would be a Clang issue!
